WcaOnRails/app/views/shared/_associated_events_picker.html.erb

```diff
@@ -12,10 +12,17 @@ Variables:
 <% disabled ||= false %>
 <% allowed_events ||= Event.official %>
 
+<% using_competition_events = events_association_name == :registration_competition_events %>
+
 <%# Note: form_builder.object is the object having associated events. %>
 <% associated_events = form_builder.object.send(events_association_name)
    all_events = allowed_events.map do |event|
-     associated_events.find_by_event_id(event.id) || associated_events.build(event_id: event.id)
+     if using_competition_events
+       competition_event = form_builder.object.competition.competition_events.find_by!(event: event)
+       associated_events.find_by_competition_event_id(event.id) || associated_events.build(competition_event: competition_event)
+     else
+       associated_events.find_by_event_id(event.id) || associated_events.build(event_id: event.id)
+     end
```

Comment:
From usage, it looks like `form_builder.object` can be either a `Registration`, `User` or `Competition`.

I'd look at moving this code out to those classes. So there'd be an `.valid_events(allowed_events)` on each class, and this whole section simplifies into

`all_events = form_builder.object.valid_events(allowed_events)`

_(The method name is probably wrong)_